### ISSUE TITLE
bug: Lifecycles events not working on IonNav componets (React)

### ISSUE THREAD

**Issue by FernetB** [ionic-team/ionic-framework#29133]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already report this problem, without success.

### Ionic Framework Version

v7.x

### Current Behavior

If I try to use the [Lifecycles](https://ionicframework.com/docs/react/lifecycle) on Nav components it does not work.

Navigating between Components inside a IonNav does not fire the lifecycle hooks.

This is frustrating because I want to handle the BackButton event on specifict components but i can't if i don't have this hooks
If i handle it in a useEffect then if i have 4 nav components mounted it will trigger the backButton in all of 4, and I only want to trigger on the current view.

### Expected Behavior

To have the Lifecycles working on the nav components

### Steps to Reproduce

Start in a page with <IonNav />
push another page that have `IonViewWillEnter and IonViewWillLeave`
then go back

You will see that neither of those hooks will trigger

### Ionic Info

Ionic:

   Ionic CLI       : 7.1.1
   Ionic Framework : @ionic/react 7.7.4

Capacitor:

   Capacitor CLI      : 5.2.3
   @capacitor/android : 5.2.3
   @capacitor/core    : 5.2.3
   @capacitor/ios     : 5.2.3

Utility:

   cordova-res : 0.15.1
   native-run  : 1.7.2

System:

   NodeJS : v18.16.0
   npm    : 9.5.1
   OS     : Windows 10

**Comment by FernetB** [comment]
Added code reproduction, but I guess if u are requesting me that, implies that lifeclycle should work on ionNav

**Comment by FernetB** [comment]
Yes, the events are working, the problem is that i don't have a simple way to know when a component pushed by a nav is active or not. With the route's lifecycles that would be easy to do, but without it is nearly impossible.
For example, if i want to handle the hardware back button on each view differently with the current tools seems impossible because the "mounted" nav components that are hidden will be also listening to the back button and this will lead unwanted behaviors. (With router lifecycles i could subscribe in the ionViewWillEnter and unsubscribe in the ionViewWillLeave)

**Comment by amandaejohnston** [comment]
If the `ionNavWillChange` event included what component is entering, would that solve things for you? Currently it doesn't emit any detail.

**Comment by amandaejohnston** [comment]
I spoke with the team on this, and there are a couple of other solutions that may work better:
1. Using `IonNav`'s existing [`getActive`](https://ionicframework.com/docs/api/nav#getactive) method to determine the active view. This could be done in your hardware back button listener, for example, to customize the logic.
2. Whenever a new component is pushed to the nav, register its hardware back button handler using a higher priority than the last (such as by using a private variable that gets incremented each time). Only the highest priority handler will be called by default. You can see more info in the hardware back button docs: https://ionicframework.com/docs/developing/hardware-back-button

Can you let me know if either of these work for your use case, and if neither of them do, why not? I want to make sure any features we add will solve your use case without needing extra unneeded APIs.

**Comment by FernetB** [comment]
I tried the first but didn't succeed because i didn't know how to do the comparition between the `getActive` view and the components to know if they were visibles or not.
The second one may work, i didn't tried it, but i found it a bit hacky. It is like i have clean and simple solutions for routing but with nav i have to do things that aren't elegant or simple, and can lead to bugs, but it is okay, if you don't want or can't adds this features i will find a workaround.
But i want to leave my opinion that routing in react is the weakest part of ionic by far, and with nav that could have change a bit if at least has the same lifecycles hooks that routing has.

**Comment by amandaejohnston** [comment]
Thank you for the details. After talking with the team about this some more, we agree that there is room to improve with the way lifecycle events are handled in `IonNav`. I'll import the ticket as a feature request since we may be adding new behavior.

Note that this any changes will also apply to Vue, since the two routing integrations have very similar code.

**Comment by FernetB** [comment]
That's awesome, thank you!
